README

U77,5 - a set of USB sound / DCF77 tools


DESCRIPTION:

This project aims to demonstrate:
a- how to build a "simple" hardware consisting of a DCF77 receiver and a
   low-cost USB soundcard.
b- evaluation of the signal from the USB sound card (using PortAudio
   see http://www.portaudio.com/)



LICENSE:

GNU LGPL (GNU Lesser Public License, see COPYING)


--- LGPL --

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser Public License
as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
See the GNU Lesser Public License for more details.

You should have received a copy of the GNU Lesser Public License
along with U77,5.  If not, see <http://www.gnu.org/licenses/>.



CHANGELOG

CHANGELOG:
==========

0.0.1
  * initial
  (2008-02-17, HA)


src/dcf77/dcf77.h

/*
 * U77,5 - a set of USB sound / DCF77 tools
 * Copyright (C) 2008 Hayati Ayguen <h_ayguen@web.de>
 * License: GNU LGPL (GNU Lesser Public License, see COPYING)
 *
 * This file is part of U77,5.
 *
 * U77,5 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * U77,5 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser Public License for more details.
 *
 * You should have received a copy of the GNU Lesser Public License
 * along with U77,5.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _U775_DCF77_H_
#define _U775_DCF77_H_

#ifndef _SVID_SOURCE
#define _SVID_SOURCE /* glibc2 needs this */
#endif
#include <time.h>
#include <stdio.h>

class DCF77
{
public:
  DCF77();
  ~DCF77();

  void initGetThreshold();
  void initGetTime();
  void newData( unsigned int framecount, const float * data );
  bool evalMinPulse(struct tm * tms, int * DCF_TZ_idx, FILE * errstream);

  typedef enum
  {
      STATE_GET_THRESH  /// initialer Zustand
    , STATE_GET_TIME
  }
    State;

  double          SampleRate;
  State            eState;
  unsigned        ChanCount;
  unsigned        ChanIdx;
  unsigned        FramesPerBuffer;
  unsigned        frameIndex;  /* Index into sample array. */

  // vars for state STATE_GET_THRESH
  double          Sum;
  volatile float Max;
  // result of state STATE_GET_THRESH
  volatile bool  ThreshStartMessage;
  volatile bool  ThreshFinishMessage;
  volatile float Mean;
  volatile float Threshold;

  // vars for state STATE_GET_TIME
  float           LastSample;
  int             FramesSinceLastPulse;
  int             LastBit;
  int             ValueMaskLo;  // DCF bits 28 .. 0
  int             ValidMaskLo;
  int             ValueMaskHi;  // DCF bits 58 .. 29
  int             ValidMaskHi;
  // result of state STATE_GET_TIME
  volatile int   FramesSinceLastMinPulse;
  volatile bool  EvaluatedMinPulse;
  volatile int   EvalValueMaskLo;  // DCF bits 28 .. 0
  volatile int   EvalValidMaskLo;
  volatile int   EvalValueMaskHi;  // DCF bits 58 .. 29
  volatile int   EvalValidMaskHi;

  volatile int   aiDiffFrames[2];
  volatile int   iDiffIdx;
  volatile int   PrintDiff;

  int            SetSysTime;
};

#endif /* _U775_DCF77_H_ */



src/tools/dcf77-settime.cpp

/*
 * U77,5 - a set of USB sound / DCF77 tools
 * Copyright (C) 2008 Hayati Ayguen <h_ayguen@web.de>
 * License: GNU LGPL (GNU Lesser Public License, see COPYING)
 *
 * This file is part of U77,5.
 *
 * U77,5 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * U77,5 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser Public License for more details.
 *
 * You should have received a copy of the GNU Lesser Public License
 * along with U77,5.  If not, see <http://www.gnu.org/licenses/>.
 */


#ifdef _MSC_VER
#include <windows.h>
#endif

#include <malloc.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <math.h>
#include <portaudio.h>

#include "../dcf77/dcf77.h"


// FramesPerBuffer (=10ms) should be the accuracy of the clock
// Evaluation for the Minute Pulse in main() can be done slower (100ms),
// as the time gap in between is accounted by FramesSinceLastMinPulse
// TODO: accounting for default latencies:
//   1- signal latency from DCF transmitter (Mainflingen near Frankfurt
//      in Germany) to place of receiption. This can be calculated using
//      the distance
//   2- latency from microphone input to processing in this application
//      This may be estimated with a direct connection from sound
//      line-out to mic-in and having a test program!
//   3- latency from starting a 3rd application / setting system time
//      to being executed.
//

// a bit annoying having to discriminate the API version
// why do they change the API?
// however: thanks for providing a usable platform independent Sound Lib
#define VER_18_1  ( ( 18 << 16 ) | 1 )
#define VER_19    ( ( 19 << 16 ) | 0 )

#ifdef _MSC_VER
  // i've installed/compiled pa_stable_v19_20071207.tar.gz
  // on MS Windows with MS Visual C++ 2003.NET
  // pa_stable_v19_20071207.tar.gz contains PortAudio Version 19.0
  #define PORTAUDIO_LIB_VERSION  VER_19
#else
  // Ubuntu Linux 6.06.1 LTS comes with PortAudio 18.1
  #define PORTAUDIO_LIB_VERSION  VER_18_1
#endif


static void
pa_error_handler (PaError pa_error)
{
  printf("PortAudio error: %s\n", Pa_GetErrorText(pa_error));

  Pa_Terminate();
  exit (1);
}


/* This routine will be called by the PortAudio engine when audio is needed.
** It may be called at interrupt level on some machines so don't do anything
** that could mess up the system like calling malloc() or free().
*/

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )

static int recordCallback(
    const void *inputBuffer
  , void *outputBuffer
  , unsigned long framesPerBuffer
  , const PaStreamCallbackTimeInfo* timeInfo
  , PaStreamCallbackFlags statusFlags
  , void *userData
  )

#else

static int recordCallback(
    void *inputBuffer
  , void *outputBuffer
  , unsigned long framesPerBuffer
  , PaTimestamp timeInfo
  , void *userData
  )

#endif
{
  DCF77 *data = (DCF77*)userData;

  data->newData( framesPerBuffer, (const float *)inputBuffer );
  return 0; // Continue
}


int main( int argc, char *argv[] )
{
  int argno;
  int ListDevices = 0;
  int DeviceCount = 0;
  int DeviceNo = 0;
  PaError pa_error;
  DCF77 data;
  double sumJitter = 0.0;
  double cntJitter = 0.0;
  const char * TZStrTab[] =
  {   "Err"
    , "MESZ (UTC+2)"
    , "MEZ (UTC+1)"
    , "Err"
  };
  const char * WeekDayStrTab[] =
  {   "Sunday"
    , "Monday"
    , "Tuesday"
    , "Wednesday"
    , "Thursday"
    , "Friday"
    , "Saturday"
  };

  printf("Pa_Initialize()\n");
  pa_error = Pa_Initialize();
  if (pa_error != paNoError)
    pa_error_handler (pa_error);
  printf("Pa_Initialize() done\n\n");


#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
  DeviceCount = Pa_GetDeviceCount();
#else
  DeviceCount = Pa_CountDevices();
#endif

  if (DeviceCount < 0 )
    pa_error_handler(DeviceCount);

  if (DeviceCount == 0)
  {
    fprintf (stderr, "No PortAudio devices found\n");
    exit(0);
  }

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
  DeviceNo = Pa_GetDefaultInputDevice(); /* default input device */;
#else
  DeviceNo = Pa_GetDefaultInputDeviceID();
#endif

  printf("Default Input Device = %d\n", DeviceNo);

  printf("\nParsing Command Lines Arguments\n");
  for ( argno = 1; argno < argc; ++argno )
  {
    if ( !strcmp(argv[argno], "--help") )
    {
      printf("%s [--help] [--list] [left|right] [44100|48000] [setsystime] [<deviceno>]\n\n", argv[0]);
    }
    else if ( !strcmp(argv[argno], "--list") )
      ListDevices = 1;
    else if ( !strcmp(argv[argno], "left") )
    {
      data.ChanIdx = 0;
      printf("Channl := Left (=0)\n");
    }
    else if ( !strcmp(argv[argno], "right") )
    {
      data.ChanIdx = 1;
      printf("Channl := Right (=1)\n");
    }
    else if ( !strcmp(argv[argno], "48000") )
    {
      data.SampleRate = atof(argv[argno]);
      data.FramesPerBuffer = 480; // = 10 * 48000 / 1000 == 10 ms
      printf("SampleRate := %f\n", data.SampleRate);
    }
    else if ( !strcmp(argv[argno], "44100") )
    {
      data.SampleRate = atof(argv[argno]);
      data.FramesPerBuffer = 441; // = 10 * 44100 / 1000 == 10 ms
      printf("SampleRate := %f\n", data.SampleRate);
    }
    else if ( !strcmp(argv[argno], "setsystime") )
    {
      data.SetSysTime = 1;
    }
    else
    {
      int tmp = atoi(argv[argno]);
      if ( tmp >= 0 && tmp < DeviceCount )
      {
        DeviceNo = tmp;
      }
      else
        fprintf(stderr, "ignoring argument '%s'! DeviceNo not in Range 0 .. %d\n", argv[argno], DeviceCount-1 );
    }
  }
  printf("End of Parsing Command Lines Arguments\n\n");

  if ( ListDevices )
  {
    printf("\n\nListing Devices\n");
    int i;
    for ( i = 0; i < DeviceCount; ++i )
    {
      const PaDeviceInfo *device_info = Pa_GetDeviceInfo(i);

      printf("%d: %s\n", i, device_info->name);
      printf("  maxInputChannels = %d\n", device_info->maxInputChannels);
#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
      printf("  default SampleRate = %f\n", device_info->defaultSampleRate);
#endif

      if (device_info->maxInputChannels < 1)
        continue;

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )

      PaStreamParameters input_param;

      memset (&input_param, sizeof(input_param), 0);

      input_param.device = i;
      input_param.channelCount = data.ChanCount = device_info->maxInputChannels;
      input_param.sampleFormat = paFloat32;
      input_param.hostApiSpecificStreamInfo = NULL;

      pa_error = Pa_IsFormatSupported(&input_param, NULL, data.SampleRate);
      if (pa_error == paFormatIsSupported)
        printf ("  supports SampleRate and Format\n");
      else
      {
        printf ("  does NOT support SampleRate and Format\n");
        printf ("  PortAudio error: %s\n", Pa_GetErrorText(pa_error));
      }

#endif


    }
    printf("End of Listing Devices\n\n\n");
  }

  do
  {
    int i = DeviceNo;
    const PaDeviceInfo *device_info = Pa_GetDeviceInfo(i);

    printf("%d: %s\n", i, device_info->name);
    printf("  maxInputChannels = %d\n", device_info->maxInputChannels);
#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
    printf("  default SampleRate = %f\n", device_info->defaultSampleRate);
#endif

    if (device_info->maxInputChannels < 1)
      break;

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
    PaStreamParameters input_param;
    memset (&input_param, sizeof(input_param), 0);

    input_param.device = i;
    input_param.channelCount = device_info->maxInputChannels;
    input_param.sampleFormat = paFloat32;
    input_param.hostApiSpecificStreamInfo = NULL;

    pa_error = Pa_IsFormatSupported(&input_param, NULL, data.SampleRate);
    if (pa_error == paFormatIsSupported)
      printf ("  supports SampleRate and Format\n");
    else
    {
      printf ("  does NOT support SampleRate and Format\n");
      printf ("  PortAudio error: %s\n", Pa_GetErrorText(pa_error));
    }
#endif
  } while (0);

  {
    PaStream*           stream;
    PaError             err = paNoError;

    data.frameIndex = 0;

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
    PaStreamParameters  inputParameters;
    inputParameters.device = DeviceNo;
    inputParameters.channelCount = data.ChanCount;
    inputParameters.sampleFormat = paFloat32;
    inputParameters.suggestedLatency = Pa_GetDeviceInfo( inputParameters.device )->defaultLowInputLatency;
    inputParameters.hostApiSpecificStreamInfo = NULL;

    /* Record some audio. -------------------------------------------- */
    err = Pa_OpenStream(
              &stream
            , &inputParameters
            , NULL                  /* &outputParameters, */
            , data.SampleRate
            , data.FramesPerBuffer
            , paClipOff      /* we won't output out of range samples so don't bother clipping them */
            , recordCallback
            , &data
            );
#else
    err = Pa_OpenStream(
              &stream
            , DeviceNo   // inputDevice
            , data.ChanCount     // numInputChannels
            , paFloat32  // inputSampleFormat
            , NULL       // void *inputDriverInfo
            , paNoDevice // outputDevice
            , 0          // numOutputChannels
            , 0          // outputSampleFormat
            , NULL       // void *outputDriverInfo
            , data.SampleRate // sampleRate
            , data.FramesPerBuffer // framesPerBuffer
            , (int)ceil(data.SampleRate * 0.100 / (double)data.FramesPerBuffer) // numberOfBuffers
            , paNoFlag   // streamFlags
            , recordCallback // PortAudioCallback *callback
            , &data      // void *userData
            );
#endif

    if( err != paNoError ) goto done;

    err = Pa_StartStream( stream );
    if( err != paNoError ) goto done;
    printf("\n\nNow recording!!\n"); fflush(stdout);

#if ( PORTAUDIO_LIB_VERSION >= VER_19 )
    while( 1 == ( err = Pa_IsStreamActive( stream ) ) )
#else
    while( 1 == ( err = Pa_StreamActive( stream ) ) )
#endif
    {
      Pa_Sleep(100);

      if ( data.ThreshStartMessage )
      {
        fprintf(stdout, "\nState STATE_GET_THRESH\n");
        fflush(stdout);
        data.ThreshStartMessage = false;
      }

      if ( data.ThreshFinishMessage )
      {
        fprintf(stdout, " => Mean = %.3f\n", data.Mean );
        fprintf(stdout, " => Threshold = %.3f\n", data.Threshold);
        fprintf(stdout, " => Max  = %.3f\n\n", data.Max );
        fflush(stdout);
        data.ThreshFinishMessage = false;
      }

      if ( data.PrintDiff )
      {
        data.PrintDiff = 0;
        if ( data.aiDiffFrames[1-data.iDiffIdx] > data.aiDiffFrames[data.iDiffIdx] )
        {
          double jitter = data.SampleRate - data.aiDiffFrames[0] - data.aiDiffFrames[1];
          if ( fabs(jitter) < 0.5*data.SampleRate )
          {
            sumJitter += jitter;
            cntJitter += 1.0;
#if 0
            double meanJitter = sumJitter / cntJitter;
            fprintf(stdout, "Pulse: %.1f %.1f\n", jitter, meanJitter );
#endif
          }
        }
      }
      if ( !data.EvaluatedMinPulse )
      {
        struct tm tms;
        int DCF_TZ_idx;
        // int Year, Month, Day, Weekday, Hour, Minute;
        if (data.evalMinPulse(&tms,&DCF_TZ_idx,stderr))
        {
          if (data.SetSysTime)
          {
            time_t tim;
            tim = mktime(&tms);
#ifdef _MSC_VER
            tms = * gmtime(&tim); // convert to UTC
            // and set to Windows system time structure
            SYSTEMTIME systime;
            systime.wYear = tms.tm_year + 1900;
            systime.wMonth = tms.tm_mon +1; // 1 .. 12
            systime.wDayOfWeek = tms.tm_wday;
            systime.wDay = tms.tm_mday;
            systime.wHour = tms.tm_hour;
            systime.wMinute = tms.tm_min;
            systime.wSecond = tms.tm_sec;
            systime.wMilliseconds = 0;
            if ( 0 == SetSystemTime(&systime) )
              fprintf(stderr, "Error setting system time\n");
#else
            if ( -1 == stime(&tim))
              fprintf(stderr, "Error setting system time: '%s'\n", strerror(errno) );
#endif
          }
          fprintf(stdout, "Date: %s, %04d-%02d-%02d  Time: %02d:%02d  %s  %f ms\n"
                        , WeekDayStrTab[tms.tm_wday], tms.tm_year + 1900, tms.tm_mon +1, tms.tm_mday
                        , tms.tm_hour, tms.tm_min
                        , TZStrTab[DCF_TZ_idx]
                        , (1000.0 * data.FramesSinceLastMinPulse / data.SampleRate)
                        );
          if (data.SetSysTime)
            goto done;
        }
        data.EvaluatedMinPulse = true;
        fflush(stdout);
      } // end if ( !data.EvaluatedMinPulse )
    } // end while( 1 == ( err = Pa_StreamActive( stream ) ) )

    if( err < 0 )
      goto done;

    err = Pa_CloseStream( stream );
    if( err != paNoError )
      goto done;

  }

done:

  printf("\n\nPa_Terminate()\n");
  Pa_Terminate();

  return 0;
}


src/dcf77/dcf77.cpp

/*
 * U77,5 - a set of USB sound / DCF77 tools
 * Copyright (C) 2008 Hayati Ayguen <h_ayguen@web.de>
 * License: GNU LGPL (GNU Lesser Public License, see COPYING)
 *
 * This file is part of U77,5.
 *
 * U77,5 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * U77,5 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser Public License for more details.
 *
 * You should have received a copy of the GNU Lesser Public License
 * along with U77,5.  If not, see <http://www.gnu.org/licenses/>.
 */


#include "dcf77.h"


DCF77::DCF77()
{
  SampleRate = 48000.0;
  eState = STATE_GET_THRESH;
  ChanCount = 1;
  ChanIdx = 0;
  FramesPerBuffer = 480; // = 10 * 48000 / 1000 == 10 ms
  frameIndex = 0;

  aiDiffFrames[0] = aiDiffFrames[1] = 0;
  iDiffIdx = 0;
  PrintDiff = 0;

  SetSysTime = 0;

  initGetThreshold();
}


DCF77::~DCF77()
{
}


void DCF77::initGetThreshold()
{
  eState = STATE_GET_THRESH;
  frameIndex = 0;
  Sum = 0.0;
  Max = -1.0F;
  ThreshStartMessage = true;
  ThreshFinishMessage = false;
}

void DCF77::initGetTime()
{
  frameIndex = 0;
  LastSample = 2.0F;
  FramesSinceLastPulse = (int)( 0.5 + 20.0 * SampleRate );
  FramesSinceLastMinPulse = -1;
  EvaluatedMinPulse = true;
  LastBit = -1;
  ValueMaskLo = 0;
  ValidMaskLo = 0;
  ValueMaskHi = 0;
  ValidMaskHi = 0;

  EvalValueMaskLo = 0;
  EvalValidMaskLo = 0;
  EvalValueMaskHi = 0;
  EvalValidMaskHi = 0;

  eState = STATE_GET_TIME;
}


bool DCF77::evalMinPulse(struct tm * tms, int * DCF_TZ_idx, FILE * errstream)
{
  // assume error
  bool retval = false;

  // Lo: 28 .. 0
  // 22222222211111111110000000000
  // 87654321098765432109876543210
  // 11111111101110000000000000000 == 0x1FF70000 == ValidMaskLo
  // 00000000100000000000000000000 ==   0x100000 == ValueMaskLo (== Startbit 20)
  // Hi: 58 .. 29
  // 555555555444444444433333333332
  // 876543210987654321098765432109
  // 111111111111111111111111111111 == 0x3FFFFFFF == ValidMaskHi

  //int DST_change    = ( (EvalValueMaskLo >> 16) & 0x01);
  int TimeZone      = ( (EvalValueMaskLo >> 17) & 0x03);

  int MinuteBCDLo   = ( (EvalValueMaskLo >> 21) & 0x0f);
  int MinuteBCDHi   = ( (EvalValueMaskLo >> 25) & 0x07);
  int ParityMinute  = ( (EvalValueMaskLo >> 21)  // 21 Start Minute
                      ^ (EvalValueMaskLo >> 22)  // 22
                      ^ (EvalValueMaskLo >> 23)  // 23
                      ^ (EvalValueMaskLo >> 24)  // 24
                      ^ (EvalValueMaskLo >> 25)  // 25
                      ^ (EvalValueMaskLo >> 26)  // 26
                      ^ (EvalValueMaskLo >> 27)  // 27 End Minute
                      ^ (EvalValueMaskLo >> 28)  // 28 Parity Minute
                      ) & 1;
  int Minute = MinuteBCDLo + 10 * MinuteBCDHi;

  int HourBCDLo     = ( (EvalValueMaskHi      ) & 0x0f);
  int HourBCDHi     = ( (EvalValueMaskHi >>  4) & 0x03);
  int ParityHour    = ( (EvalValueMaskHi)        // 29 Start Hour
                      ^ (EvalValueMaskHi >>  1)  // 30
                      ^ (EvalValueMaskHi >>  2)  // 31
                      ^ (EvalValueMaskHi >>  3)  // 32
                      ^ (EvalValueMaskHi >>  4)  // 33
                      ^ (EvalValueMaskHi >>  5)  // 34 End Hour
                      ^ (EvalValueMaskHi >>  6)  // 35 Parity Hour
                      ) & 1;
  int Hour = HourBCDLo + 10 * HourBCDHi;

  int DayBCDLo      = ( (EvalValueMaskHi >>  7) & 0x0f);
  int DayBCDHi      = ( (EvalValueMaskHi >> 11) & 0x03);
  int Weekday       = ( (EvalValueMaskHi >> 13) & 0x07);  // 1 .. 7
  int MonthBCDLo    = ( (EvalValueMaskHi >> 16) & 0x0f);
  int MonthBCDHi    = ( (EvalValueMaskHi >> 20) & 0x01);
  int YearBCDLo     = ( (EvalValueMaskHi >> 21) & 0x0f);
  int YearBCDHi     = ( (EvalValueMaskHi >> 25) & 0x0f);

  int Day           = DayBCDLo    + 10 * DayBCDHi;
  int Month         = MonthBCDLo  + 10 * MonthBCDHi;
  int Century       = 20;
  int Year          = YearBCDLo   + 10 * YearBCDHi;
  int ParityDate    = ( (EvalValueMaskHi >>  7)      // 36 Start Day
                      ^ (EvalValueMaskHi >>  8)      // 37
                      ^ (EvalValueMaskHi >>  9)      // 38
                      ^ (EvalValueMaskHi >> 10)      // 39
                      ^ (EvalValueMaskHi >> 11)      // 40
                      ^ (EvalValueMaskHi >> 12)      // 41 End Day
                      ^ (EvalValueMaskHi >> 13)      // 42 Start Weekday
                      ^ (EvalValueMaskHi >> 14)      // 43
                      ^ (EvalValueMaskHi >> 15)      // 44 End Weekday
                      ^ (EvalValueMaskHi >> 16)      // 45 Start Month
                      ^ (EvalValueMaskHi >> 17)      // 46
                      ^ (EvalValueMaskHi >> 18)      // 47
                      ^ (EvalValueMaskHi >> 19)      // 48
                      ^ (EvalValueMaskHi >> 20)      // 49 End Month
                      ^ (EvalValueMaskHi >> 21)      // 50 Start Year
                      ^ (EvalValueMaskHi >> 22)      // 51
                      ^ (EvalValueMaskHi >> 23)      // 52
                      ^ (EvalValueMaskHi >> 24)      // 53
                      ^ (EvalValueMaskHi >> 25)      // 54
                      ^ (EvalValueMaskHi >> 26)      // 55
                      ^ (EvalValueMaskHi >> 27)      // 56
                      ^ (EvalValueMaskHi >> 28)      // 57 End Year
                      ^ (EvalValueMaskHi >> 29)      // 58 Parity Year
                      ) & 1;

  if ( (EvalValidMaskLo & 0x1FF70000) != 0x1FF70000
    || (EvalValidMaskHi & 0x3FFFFFFF) != 0x3FFFFFFF )
  {
    /* fprintf(errstream, "Error: Not enough bits collected\n"); */
  }
  else if ( (EvalValueMaskLo &   0x100000) != 0x100000 )
  {
    if (errstream)
      fprintf(errstream, "Error: Startbit 20 not set\n");
  }
  else if ( 0 == TimeZone || 3 == TimeZone )
  {
    if (errstream)
      fprintf(errstream, "Error: TimeZone neither MEZ nor MESZ\n");
  }
  else if ( 0 != ParityMinute )
  {
    if (errstream)
      fprintf(errstream, "Error: Defect Parity for Minute\n");
  }
  else if ( MinuteBCDLo < 0 || MinuteBCDLo > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Lower digit for Minute not in range 0 .. 9\n");
  }
  else if ( MinuteBCDHi < 0 || MinuteBCDHi >= 6 )
  {
    if (errstream)
      fprintf(errstream, "Error: Higher digit for Minute not in range 0 .. 5\n");
  }
  else if ( 0 != ParityHour )
  {
    if (errstream)
      fprintf(errstream, "Error: Defect Parity for Hour\n");
  }
  else if ( HourBCDLo < 0 || HourBCDLo > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Lower digit for Hour not in range 0 .. 9\n");
  }
  else if ( HourBCDHi < 0 || HourBCDHi >= 3 )
  {
    if (errstream)
      fprintf(errstream, "Error: Higher digit for Hour not in range 0 .. 2\n");
  }
  else if ( Hour >= 24 )
  {
    if (errstream)
      fprintf(errstream, "Error: Hour not in range 0 .. 23\n");
  }

  else if ( 0 != ParityDate )
  {
    if (errstream)
      fprintf(errstream, "Error: Defect Parity for Date\n");
  }
  else if ( DayBCDLo < 0 || DayBCDLo > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Lower digit for Day not in range 0 .. 9\n");
  }
  else if ( DayBCDHi < 0 || DayBCDHi > 3 )
  {
    if (errstream)
      fprintf(errstream, "Error: Higher digit for Day not in range 0 .. 2\n");
  }
  else if ( Day < 1 || Day > 31 )
  {
    if (errstream)
      fprintf(errstream, "Error: Day not in range 1 .. 31\n");
  }
  else if ( Weekday < 1 || Weekday > 7 )
  {
    if (errstream)
      fprintf(errstream, "Error: Weekday not in range 1 .. 7\n");
  }
  else if ( MonthBCDLo < 0 || MonthBCDLo > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Lower digit for Month not in range 0 .. 9\n");
  }
  else if ( MonthBCDHi < 0 || MonthBCDHi > 1 )
  {
    if (errstream)
      fprintf(errstream, "Error: Higher digit for Month not in range 0 .. 1\n");
  }
  else if ( Month < 1 || Month > 12 )
  {
    if (errstream)
      fprintf(errstream, "Error: Month not in range 1 .. 12\n");
  }
  else if ( YearBCDLo < 0 || YearBCDLo > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Lower digit for Year not in range 0 .. 9\n");
  }
  else if ( YearBCDHi < 0 || YearBCDHi > 9 )
  {
    if (errstream)
      fprintf(errstream, "Error: Higher digit for Year not in range 0 .. 9\n");
  }
  else
  {
    tms->tm_sec   = 0; // 0 .. 59, 60 for leap sec
    tms->tm_min   = Minute; // 0 .. 59
    tms->tm_hour  = Hour; // 0 .. 23
    tms->tm_mday  = Day;  // 1 .. 31
    tms->tm_mon   = Month -1; // 0 .. 11
    tms->tm_year  = Century * 100 + Year - 1900; // years since 1900
    // tms->tm_wday  = 0; // mktime() ignores this
    tms->tm_wday  = (Weekday == 7) ? 0 : Weekday; // convert from 1..7 to 0..6
    tms->tm_yday  = 0; // mktime() ignores this
    tms->tm_isdst = (1 == TimeZone) ? 1 : 0; //positive if daylight saving time is in effect, zero if it is not, and negative if unknown
    *DCF_TZ_idx   = TimeZone;
    retval = true;
  }

  return retval;
}


void DCF77::newData( unsigned int framecount, const float * data )
{
  unsigned int i;
  unsigned int idx;
  bool  ReSync;
  float LocalLastSample;

  switch( eState )
  {
    case STATE_GET_THRESH:
      idx = ChanIdx;
      for ( i = 0; i < framecount; ++i, idx += ChanCount )
      {
        // gather statistics
        if ( data[idx] > Max )
          Max = data[idx];
        Sum += data[idx];
      }
      frameIndex += framecount;

      // evaluate statistics after 10 seconds
      if ( (double)frameIndex >= 10.0 * SampleRate )
      {
        const double dMean = Sum / frameIndex;
        Mean      = (float)dMean;
        // Signal Power should not exceed 7/10 th of Mean to Max voltage
        Threshold = (float)( dMean + 0.7 * ( Max - dMean ) );
        ThreshFinishMessage = true;
        // State finished --> next state := STATE_GET_TIME
        initGetTime();
      }
      break;

    case STATE_GET_TIME:

      LocalLastSample = LastSample;
      idx = ChanIdx;
      ReSync = false;

      if ( FramesSinceLastMinPulse >= 0 )
        FramesSinceLastMinPulse += framecount;

      for ( i = 0; i < framecount; ++i, idx += ChanCount )
      {
        if ( LocalLastSample < Threshold && data[idx] >= Threshold )
        {
          const float MSecsSinceLastPulse = (float)( ( FramesSinceLastPulse + i ) * 1000.0 / SampleRate );
          if      ( ( -1 == LastBit && MSecsSinceLastPulse >  60.0 && MSecsSinceLastPulse < 140.0 )  // ~ 100 ms
                  ||( -1 == LastBit && MSecsSinceLastPulse > 160.0 && MSecsSinceLastPulse < 240.0 )  // ~ 200 ms
                  )
          {
            LastBit = ( MSecsSinceLastPulse < 150.0 ) ? 0 : 1;
            // DCF bits 28 .. 0: add 1 new bit from Hi and shift old bits
            ValueMaskLo = ( (ValueMaskHi & 1) << 28 ) | ( ValueMaskLo >> 1 );
            ValidMaskLo = ( (ValidMaskHi & 1) << 28 ) | ( ValidMaskLo >> 1 );
            // DCF bits 58 .. 29 == 29 .. 0: add 1 new bit and shift old bits
            ValueMaskHi = ( LastBit << 29 ) | ( ValueMaskHi >> 1 );
            ValidMaskHi =       ( 1 << 29 ) | ( ValidMaskHi >> 1 );

            aiDiffFrames[iDiffIdx] = FramesSinceLastPulse + i;
            iDiffIdx = 1 - iDiffIdx;
            PrintDiff = 1;
            FramesSinceLastPulse = - (int)i;
          }
          else if ( ( 0 == LastBit && MSecsSinceLastPulse > 860.0 && MSecsSinceLastPulse < 940.0 )  // ~ 900 ms
                  ||( 1 == LastBit && MSecsSinceLastPulse > 760.0 && MSecsSinceLastPulse < 840.0 )  // ~ 800 ms
                  )
          {
            LastBit = -1;   // after 100 ms or 200 ms Pulse at Second pulse

            aiDiffFrames[iDiffIdx] = FramesSinceLastPulse + i;
            iDiffIdx = 1 - iDiffIdx;
            PrintDiff = 1;
            FramesSinceLastPulse = - (int)i;
          }
          else if ( ( 0 == LastBit && MSecsSinceLastPulse > 1860.0 && MSecsSinceLastPulse < 1940.0 )  // ~ 1900 ms
                  ||( 1 == LastBit && MSecsSinceLastPulse > 1760.0 && MSecsSinceLastPulse < 1840.0 )  // ~ 1800 ms
                  )
          {
            LastBit = -1; // after 100 ms or 200 ms Pulse at Minute pulse
            FramesSinceLastMinPulse = framecount - i;
            EvalValueMaskLo = ValueMaskLo;
            EvalValidMaskLo = ValidMaskLo;
            EvalValueMaskHi = ValueMaskHi;
            EvalValidMaskHi = ValidMaskHi;
            EvaluatedMinPulse = false;

            aiDiffFrames[iDiffIdx] = FramesSinceLastPulse + i;
            iDiffIdx = 1 - iDiffIdx;
            PrintDiff = 1;
            FramesSinceLastPulse = - (int)i;
          }
          else if ( MSecsSinceLastPulse >= 20000.0 && MSecsSinceLastPulse < 50000.0 )  // initial pulse search?
          {
            LastBit = -1;   // ignore
            FramesSinceLastPulse = - (int)i;
          }
          else if ( -1 == LastBit && MSecsSinceLastPulse < 30.0 )
          {
            // filter noise!
            LastBit = -1;
            //fprintf(stderr, "ignore after %f ms\n", MSecsSinceLastPulse);
            // do not set FramesSinceLastPulse !!!
          }
          else
          {
            //fprintf(stderr, "resync: with LastBit=%d after %f ms\n", LastBit, MSecsSinceLastPulse);
            ReSync = true;  // sync error!
            LastBit = -1;
            FramesSinceLastPulse = - (int)i;
          }
        }
        LocalLastSample = data[idx];
      } // end for
      FramesSinceLastPulse += framecount;
      LastSample = LocalLastSample;

      if ( ( FramesSinceLastPulse > 10.0 * SampleRate
          && FramesSinceLastPulse < 20.0 * SampleRate )
          || ReSync
          )
      {
        initGetThreshold();
      }
      break;

    default:
      ;
  }
}

